tools-local/Microsoft.ML.InternalCodeAnalyzer/BestFriendAnalyzer.cs

```diff
@@ -104,5 +107,18 @@ private void Analyze(SemanticModelAnalysisContext context)
                 context.ReportDiagnostic(diagnostic);
             }
         }
+
+        private void Analyze(SemanticModelAnalysisContext context)
+        {
+            // Analyzer function for BestFriend defined in Microsoft.ML
+            _attributeName = "Microsoft.ML.BestFriendAttribute";
+            _assemblyAttributeName = "Microsoft.ML.WantsToBeBestFriendsAttribute";
+            AnalyzeCore(context, _attributeName, _assemblyAttributeName);
+
+            // Analyzer function for BestFriend defined in Microsoft.ML.Hidden
+            _attributeName = "Microsoft.ML.Hidden.BestFriendAttribute";
```

Comment:
I don't think mutating the object is a great design/approach. Can we get away with not having these fields? Or possibly following the same approach as the ContractsCheckAnalyzer?

[https://github.com/dotnet/machinelearning/blob/65c3c7c14b5da2655cce2c393a5b85546e5091fc/tools-local/Microsoft.ML.InternalCodeAnalyzer/ContractsCheckAnalyzer.cs#L199-L200](https://github.com/dotnet/machinelearning/blob/65c3c7c14b5da2655cce2c393a5b85546e5091fc/tools-local/Microsoft.ML.InternalCodeAnalyzer/ContractsCheckAnalyzer.cs#L199-L200)

Reply:
Ok. Now this class is stateless.

---
In reply to: [236733064](https://github.com/dotnet/machinelearning/pull/1724#discussion_r236733064) [](ancestors = 236733064)